codegenerator/cli/templates/static/codegen/src/EventRouter.res

```diff
@@ -0,0 +1,109 @@
+open Belt
+
+exception EventDuplicate
+exception WildcardCollision
+
+module Group = {
+  type t<'a> = {
+    mutable wildcard: option<'a>,
+    all: array<'a>,
+    byContractName: dict<'a>,
+  }
+
+  let empty = () => {
+    wildcard: None,
+    all: [],
+    byContractName: Js.Dict.empty(),
+  }
+
+  let addOrThrow = (group: t<'a>, event, ~contractName, ~isWildcard) => {
+    let {all, byContractName, wildcard} = group
+    switch byContractName->Utils.Dict.dangerouslyGetNonOption(contractName) {
+    | Some(_) => raise(EventDuplicate)
+    | None =>
+      if isWildcard && wildcard->Option.isSome {
+        raise(WildcardCollision)
+      } else {
+        if isWildcard {
+          group.wildcard = Some(event)
+        }
+        all->Js.Array2.push(event)->ignore
+        byContractName->Js.Dict.set(contractName, event)
+      }
+    }
+  }
+
+  let get = (group: t<'a>, ~contractAddress, ~contractAddressMapping) =>
+    switch group {
+    | {all: [event]} => Some(event)
+    | {wildcard, byContractName} =>
+      switch contractAddressMapping->ContractAddressingMap.getContractNameFromAddress(
+        ~contractAddress,
+      ) {
+      | Some(contractName) => byContractName->Utils.Dict.dangerouslyGetNonOption(contractName)
+      | None => wildcard
+      }
+    }
+}
+
+type t<'a> = dict<Group.t<'a>>
```

Comment:
Pretty much the same, but made it generic and renamed to Router instead of Lookup

Reply:
Cool, haha I actually had it as generic initially but then there was only one use of it so made it a static type.

Reply:
Bit of a pain that I can't see the diff so easily.